### PR TITLE
Select 05

### LINKED ISSUE #5: Add Form Widgets
@smhigley commented on [Fri Dec 09 2016](https://github.com/dojo/widget-core/issues/174)

**Enhancement** 

Related to #2. Update existing textinput and button components to be accessible, and add the following basic form widgets based on #85:

- [x] Email/number/password/etc to existing textInput
- [x] Checkbox
- [x] Radio buttons
- [x] Select box
- [x] Textarea

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Resolves #5, adds the last form widget

Adds a select widget with native and non-native options. In addition to the usual form properties, it has the following API:
```
multiple?: boolean;
options?: SelectOption[];
renderOption?(option: SelectOption): DNode;
useNativeSelect?: boolean;
onBlur?(event: FocusEvent): void;
onChange?(option: SelectOption): void;
onClick?(event: MouseEvent): void;
onFocus?(event: FocusEvent): void;
onKeyDown?(event: KeyboardEvent): void;
```

A sample native select could be rendered as follows:
```
const options = [
  { label: 'First', value: 'one' },
  { label: 'Second', value: 'two' }
];

w(Select, {
  options: options,
  useNativeSelect: true,
  value: this.state.selectValue,
  onChange: (option: SelectOption) => this.setState({ selectValue: option.value })
});
```

A non-native select with custom option elements would use `renderOption`:
```
w(Select, {
  options: options,
  renderOption: (option: SelectOption) => v('div', { classes: this.classes(customCss.testClass) }, [ option.label ]),
  value: this.state.selectValue,
  onChange: (option: SelectOption) => this.setState({ selectValue: option.value })
});
```

### CODE REVIEW

File: src/select/Select.ts [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+import uuid from '@dojo/core/uuid';
+import Label, { LabelOptions } from '../label/Label';
+import * as css from './styles/select.m.css';
+
+/**
+ * @type SelectOption
+ *
+ * Properties that can be set on a Select component
+ *
+ * @property label				Text to display to the user
```
Comment (bitpshr) [inline src/select/Select.ts]: TypeDoc should be indented using spaces now for consistency.

File: src/select/Select.ts [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+	private _onBlur (event: FocusEvent) { this.properties.onBlur && this.properties.onBlur(event); }
+	private _onClick (event: MouseEvent) { this.properties.onClick && this.properties.onClick(event); }
+	private _onFocus (event: FocusEvent) { this.properties.onFocus && this.properties.onFocus(event); }
+	private _onKeyDown (event: KeyboardEvent) { this.properties.onKeyDown && this.properties.onKeyDown(event); }
+
+	// native select events
+	private _onNativeChange (event: Event) {
+		const { options = [] } = this.properties;
+		const option = options.filter((option: SelectOption) => option.value === (<HTMLInputElement> event.target).value)[0];
```
Comment (bitpshr) [inline src/select/Select.ts]: I'd break `value` out into its own variable so you don't have to dig into `event` every iteration of `filter`. I'd also use `find` instead of `filter` since you're always using the first item that satisfies the filter condition:

```ts
const value = (<HTMLInputElement> event.target).value;
const option = options.find((option: SelectOption) => option.value === value);
```

File: src/select/Select.ts [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+		this.properties.onChange && this.properties.onChange(option);
```
Comment (bitpshr) [inline src/select/Select.ts]: You can destructure `onChange`.

File: src/select/Select.ts [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+
+		const optionNodes = [];
+		let optionNode, option;
+		for (let i = 0; i < options.length; i++) {
+			option = options[i];
+			option.id = option.id || uuid() + '';
+			option.selected = multiple ? option.selected : value === option.value;
+			optionNode = renderOption ? renderOption(option) : option.label;
+
+			optionNodes.push(v('div', {
```
Comment (bitpshr) [inline src/select/Select.ts]: Did you decide to not use the registry approach to allow custom option components?

File: src/select/Select.ts [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+			invalid ? css.invalid : null,
+			invalid === false ? css.valid : null,
+			multiple ? css.multiselect : null,
+			readOnly ? css.readonly : null,
+			required ? css.required : null
+		];
+
+		let rootWidget;
+
+		const select = useNativeSelect ? this.renderNativeSelect() : multiple ? this.renderCustomMultiSelect() : this.renderCustomSelect();
```
Comment (bitpshr) [inline src/select/Select.ts]: Double ternary statements are hard to read.

File: src/select/Select.ts [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+			value
+		} = this.properties;
+
+		const {
+			_open,
+			_focusedIndex,
+			_selectId
+		} = this;
+
+		const selectedOption = options.filter((option: SelectOption) => option.value === value)[0] || options[0];
```
Comment (bitpshr) [inline src/select/Select.ts]: `Array.first` instead of `Array.filter`

File: src/select/Select.ts [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+			case keys.enter:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.space:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.escape:
+				this._closeSelect();
+				break;
+			case keys.down:
```
Comment (bitpshr) [inline src/select/Select.ts]: Can a `disabled` option be focused using the keyboard? I didn't see any logic to find the next enabled option if the currently-focused option is disabled. Curious because I had to do something like that for TabPane.

File: src/select/styles/example.m.css [hunk 8]
```diff
@@ -0,0 +1,28 @@
+@import '../../common/styles/variables.css';
+
+.option {
```
Comment (bitpshr) [inline src/select/styles/example.m.css]: I think most of the example page CSS can go away. Tom is pulling out theme styles from structural styles at dojo/widgets#77. These aesthetic styles could go into the theme CSS, and any minor page styling could just be done inline. I think we should try to keep `styles` clean of example-specific stuff.

Comment (smhigley) [inline src/select/styles/example.m.css]: Oops, I actually put the example styling inline, and forgot to delete this :P

File: src/select/Select.ts (hunk 4 above)
Comment (smhigley) [inline src/select/Select.ts]: Yeah, in the end it didn't seem necessary, and I didn't really want to split out custom option nodes into a separate widget. I'm open to hearing reasons to do so, though

File: src/select/Select.ts (hunk 7 above)
Comment (smhigley) [inline src/select/Select.ts]: Yes, disabled options should be (and are) focusable using the keyboard: https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_disabled_controls

File: src/select/Select.ts (hunk 2 above)
Comment (smhigley) [inline src/select/Select.ts]: Quick question -- IE supports `Array.filter` but not `Array.find`. Should I still use find, and is there a dojo helper or something I should use to get IE11 support? (I didn't see anything in `dojo/core/lang` or `dojo/core/util`)

Comment (bitpshr) [inline src/select/Select.ts]: We have a shim in `@dojo/shim/array`.

File: src/select/Select.ts [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+	}
+
+	// custom select events
+	private _openSelect() {
+		this._open = true;
+		this._ignoreBlur = false;
+		this._focusedIndex = this._focusedIndex || 0;
+		this.invalidate();
+	}
+	private _closeSelect() {
```
Comment (bitpshr) [inline src/select/Select.ts]: New line

File: src/select/Select.ts [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+			name,
+			options = [],
+			readOnly,
+			required,
+			value
+		} = this.properties;
+
+		/* create option nodes */
+		const optionNodes = [];
+		for (let option of options) {
```
Comment (bitpshr) [inline src/select/Select.ts]: `map` may be cleaner

**Comment by smhigley** [comment]
Waiting for https://github.com/dojo/widget-core/issues/442 to fix test issue causing build failure.

File: src/select/Select.ts [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+	value?: string;
+	onBlur?(event: FocusEvent): void;
+	onChange?(option: OptionData): void;
+	onClick?(event: MouseEvent): void;
+	onFocus?(event: FocusEvent): void;
+	onKeyDown?(event: KeyboardEvent): void;
+}
+
+// This should have a lookup somewhere
+const keys = {
```
Comment (bitpshr) [inline src/select/Select.ts]: I wonder if it's time to pull this out into `common/keys` or something similar.

File: src/select/Select.ts (hunk 7 above)
Comment (bitpshr) [inline src/select/Select.ts]: I need to update ComboBox in that case. Thanks.

File: src/select/Select.ts [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+		const {
+			describedBy,
+			disabled,
+			invalid,
+			options = [],
+			readOnly,
+			required
+		} = this.properties;
+
+		const {
```
Comment (bitpshr) [inline src/select/Select.ts]: I think you can combine this destructuring:

```ts
const {
	_focusedIndex,
	properties: {
		describedBy,
		disabled,
		invalid,
		options = [],
		readOnly,
		required
	}
} = this;
```

Comment (smhigley) [inline src/select/Select.ts]: Ohhh, nice! I didn't even think about nested destructuring.

File: src/select/Select.ts (hunk 11 above)
Comment (smhigley) [inline src/select/Select.ts]: Probably. I think it might be better to do it in a separate PR, though, and change it in ComboBox, Tabs, and wherever else as well.

File: src/select/Select.ts (hunk 7 above)
Comment (smhigley) [inline src/select/Select.ts]: Cool 👍. It's not universal (i.e. not every single disabled focusable thing should continue to get focus), but it's one of the exceptions outlined in the spec. I think Combobox could probably have focusable disabled options too, though :)

**Comment by rishson** [comment]
@smhigley do you want to change `useNativeSelect` to `useNativeElement` in this PR?

File: src/select/Select.ts [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+			v('span', {
+				classes: this.classes(css.arrow)
+			})
+		]);
+	}
+
+	renderCustomMultiSelect(): DNode {
+		const {
+			_focusedIndex,
+			properties = {
```
Comment (rishson) [inline src/select/Select.ts]: causing tslint 'unused variable' error.
